### PR TITLE
Contracts holding NFTs

### PR DESCRIPTION
**Pull request by AlexanderLimonov**
Prototype of changes to NFT functionality enabling contracts (not packages) to control tokens (note that this is not meant to be merged, hence it's a draft PR - in particular, there are no unit tests, but I can provide some primitive scripts to validate basic functionality)

### CODE REVIEW

**Comment by AlexanderLimonov** [comment]
Necessary changes:
- check recipients for addressability (Account or Hash Keys)

Possible improvements:
- pass caller directly to functions instead of checking against a runtime arg

Unit tests:
- check that present manner of extracting caller (which gets you the next to last element in the call stack) gets the intended caller
- check that a recipient contract cannot re-transfer a token twice
- no double minting (probable bug here, likely unrelated to the PR)

File: cep47-logic/src/lib.rs [hunk 1]
```diff
@@ -15,6 +16,34 @@ pub mod tests;
...
+    }
+
+    fn matches_caller(&self, caller: &CallStackElement) -> bool {
+        if !self.is_addressable() {
+            return false;
+        }
+        match (self, caller) {
+            (Key::Account(account_hash),  CallStackElement::Session { account_hash: account_hash_caller}) =>
+                account_hash == account_hash_caller,
+            (Key::Hash(hash_addr), CallStackElement::StoredContract { contract_package_hash: _, contract_hash: contract_hash_addr_caller}) =>
```
Comment (zie1ony) [inline cep47-logic/src/lib.rs:40]: ContractPackageHash should be extracted here, not ContractHash.

Comment (zie1ony) [reply cep47-logic/src/lib.rs:40]: I've also experimented with this feature and my version looks like that:
```
fn get_caller_id() -> Key {
    match runtime::get_call_stack().first().unwrap_or_revert() { // first or last?
        CallStackElement::Session { account_hash } => 
            account_hash_to_id(account_hash),
        CallStackElement::StoredSession { account_hash, contract_package_hash: _, contract_hash: _ } => // can stored contract call stored session? 
            account_hash_to_id(account_hash),
        CallStackElement::StoredContract { contract_package_hash, contract_hash: _ } =>
            contract_package_hash_to_id(contract_package_hash)
    }
}

fn account_hash_to_id(account_hash: &AccountHash) -> Key {
    (*account_hash).into()
}

fn contract_package_hash_to_id(hash: &ContractPackageHash) -> Key {
    (*hash).into()
}
```

Comment (AlexanderLimonov) [reply cep47-logic/src/lib.rs:40]: If tokens are to be controlled by a contract, it seems like it should be some definite piece of code, not just a pointer to a collection of contracts?

In your example, by the way, I believe the recovery of the calling entity is incorrect - as far as @goral09 and I could tell, it should be the next-to-last element in the stack, not first (or last - that would be the NFT contract itself, I think).

**Comment by osmankahraman** [comment]
Nice

**Pull request closed without merging**